### ISSUE TITLE
Large input weights fluctuations for different king positions

### ISSUE THREAD

**Issue by ddobbelaere** [glinscott/nnue-pytorch#53]
I'm trying to get some more grip onto the input layer and decided to change the order of piece and king indices (along with undoing the king index flipping, in order to not let my brain explode in interpreting the plots). Here's the diff w.r.t. master:

```
--- a/visualize.py
+++ b/visualize.py

-                inpos = [(7-kipos[0])+pipos[0]*8,
-                         kipos[1]+(7-pipos[1])*8]
-                d = - 8 if piece < 2 else 48 + (piece // 2 - 1) * 64
+                inpos = [8*kipos[0]+pipos[0],
+                         8*(7-kipos[1])+(7-pipos[1])]
+                d = -2*(7-kipos[1]) - 1 if piece < 2 else 48 + \
+                    (piece // 2 - 1) * 64
```

The plots become:

![reordered nn-62ef826d1a6d nnue_input-weights](https://user-images.githubusercontent.com/38748663/106363185-5d974080-6327-11eb-8ae5-fde6a12bd7ac.jpg)
![reordered epoch=427 ckpt_input-weights](https://user-images.githubusercontent.com/38748663/106363187-5f610400-6327-11eb-978c-395848a5c0c0.jpg)

Some strange things/observations:
- ~There is something special going on with king on first rank for the queen not on first rank weights, both in master and in vondele net. Example hidden input neuron (there are a lot more of them):~ BUG in code, forgot to update offset...

- Look at this weird pattern in vondele net:
![image](https://user-images.githubusercontent.com/38748663/106363184-5839f600-6327-11eb-9770-973ca71bf899.png)
The only explanation I can give for it is that the hidden input neuron serves the purpose of detecting two "hidden features" at once. If the king is on the back rank (or on a6, a7, d7, h6 or h7), something completely different is happening/detected. Either that, or some gap in training data?

**Comment by vondele** [comment]
this reminds me of your correlation plots, which has this strange pattern for my nets https://github.com/glinscott/nnue-pytorch/issues/45. 

The other thing is, in pytorch we are skipping all fens where we are in check (or the bestmove is a capture).. could these features correspond to e.g. being in check?

**Comment by ddobbelaere** [comment]
> The other thing is, in pytorch we are skipping all fens where we are in check (or the bestmove is a capture).. could these features correspond to e.g. being in check?

Hmm, not really I think. The "in check" weights are the smaller patterns in each 8x8 subsquare that represent one king position.

**Comment by ddobbelaere** [comment]
Turns out I had a bug in the code (forgot to update the offset...). At least the "king on first rank" and weird queen weights issue is gone :)

**Comment by ddobbelaere** [comment]
The bias term of the weirdly shaped neuron is -0.94, so I was actually amazed that those "comb-like" back-rank weights were still alive and updated (almost all negative and only very small positive *for opposite queens*). I just thought they were dead , which would be pretty bad, and compared checkpoint 245 and 427, but they are not dead!!

Turning off coalescing and only plotting real features yields a different picture:

![image](https://user-images.githubusercontent.com/38748663/106368239-782de180-6348-11eb-9b5a-b479768e1542.png)

Note that now, for king on backrank (or a6, a7, d7, h6 and h7) *all weights* are negative. Now guess what the bias is? -0.60.
This neuron is as dead as could be for those king positions and should never ever change in training (grad is identical to zero w.r.t those weights), which is weirdly not what is happening (the weights change only a tiny bit). Probably some weird Ranger artifact...

I don't like this, I think this "comb" should have been part of a feature recognizer but "never made the jump" once weights started to go positive for other king positions. OR I am completely missing something here, please tell me so...

**Comment by ddobbelaere** [comment]
Here's the plot with virtual features only:

![image](https://user-images.githubusercontent.com/38748663/106368737-3dc64380-634c-11eb-92a3-264081649567.png)

Let's further factorize, HalfK only:

![image](https://user-images.githubusercontent.com/38748663/106368929-b8dc2980-634d-11eb-9e3f-074cd71de2bb.png)

P only:

![image](https://user-images.githubusercontent.com/38748663/106368969-06589680-634e-11eb-88e4-a3c20b6e9c90.png)

**Comment by ddobbelaere** [comment]
Ok, I think the "comb" weights are still changing during training because something called "gradient centralization" is used (default on in Ranger). Even if the gradient w.r.t. those weights is zero (which I am 99% sure, as too much weights are negative and only opposite queens slightly positive), the update term still has a contribution in the direction of "the mean of other gradients", see equation (1) of https://arxiv.org/pdf/2004.01461v2...

I think, if gradient centralization would be off, those weights would stop updating during training.

**Comment by ddobbelaere** [comment]
There's another input neuron with this behavior. In "reordered epoch=427.cpkt" plot above, third row, 7 from the right, there are seemingly anomalous blue spots for king positions a5, a7 and h7, combined with "a sea of red for opponent queen". I'll check the bias term later...

**Comment by ddobbelaere** [comment]
I think this issue occurs because of our sparse excitation of the input layer. If this would be an image recognition problem, the problem would resolve itself as concurrent activations of king positions from both "disjunct" regions would fuse the weights back to each other. Now I think some neuron are effectively dead for some king positions. You can say that the input features are partitioned into 64 disjunct sets, one for each king positions, that happen to share the same neurons (and bias terms).

I did a quick literature search on training of "sparsely excited FC layers" but did not find anything interesting.

**Comment by ddobbelaere** [comment]
There was some hope that the buggy initialization of virtual features (fixed in #56) was the root cause of this issue. @vondele has shared a net from his run96 with new initialization. However, the same observations are apparent.

![reordered nn-run96-epoch353 nnue_input-weights](https://user-images.githubusercontent.com/38748663/106925720-a3ca1680-6710-11eb-9736-266e40a8e2e0.jpg)
![image](https://user-images.githubusercontent.com/38748663/106925489-6e252d80-6710-11eb-89a0-9104bdc4a1ff.png)
![image](https://user-images.githubusercontent.com/38748663/106925568-8006d080-6710-11eb-8946-df1d682f7241.png)
![image](https://user-images.githubusercontent.com/38748663/106925681-9876eb00-6710-11eb-8e27-2e22319aef44.png)

**Comment by ddobbelaere** [comment]
Btw @sergiovieri mentioned this on Discord:

> Btw, when using ranger vs sgd, the net produced looks very different
> Using ranger seems to make the net have very different weights for different king positions (same piece&position).

**Comment by sergiovieri** [comment]
The weird halfK weights is probably because not scaling the gradients of the halfK feature down (which the c++ trainer does).

I think the reason for the scaling down is that we want to have the same gradients as when the input is normalized (that is, when all features are scaled to 0-1). Meanwhile, we cannot just scale down the input since the propagated value is dependent on the number of activated input.

**Comment by vondele** [comment]
@sergiovieri do you have a patch that correctly does this for the pytorch trainer?

**Comment by sergiovieri** [comment]
@vondele currently just a hardcoded fix here https://github.com/sergiovieri/nnue-pytorch/blob/e14671df311c56f8a09be06a633ecce1f159960f/model.py#L120-L125

Hardcoded to /= 30, but it should be /= active feature count (between 1-30). The gradient is sometimes too small, but should be better than constantly having 30x too big. You can try whether this have any effect when using ranger.

**Comment by Sopel97** [comment]
to match the nodchip trainer the gradient would have to be divided by `input^2`.

**Comment by vondele** [comment]
I'll give scaled gradient a go with 10x reduced...that's already a big change compared to the current implementation, I would guess.

**Comment by vondele** [comment]
no success with 10x reduced gradient.. looks very similar to me.

**Comment by ddobbelaere** [comment]
@vondele Hmm, ok. Do you mean there are still large differences between king positions (for a given neuron) or just that the elo performance is about the same?

Given that is seems to happen more often for king positions on back rank, how certain are we that those positions get excited a lot? How likely is it that, for a given batch size (let's say of 8192) and a given set of training data, that king position is on the back rank?

**Comment by vondele** [comment]
I looked at Elo performance only. The best net of the series would be https://drive.google.com/file/d/1It3z_QQjmEijBgnVKjSIWHWwRCVTry97/view?usp=sharing (run98/run2/nn-epoch210.nnue)

I've now started an optimization with the factorizer off (so HalfKP not HalfKP^)...

**Comment by ddobbelaere** [comment]
Indeed no improvement for this issue.

![reordered nn-run98run2-epoch210 nnue_input-weights](https://user-images.githubusercontent.com/38748663/107118190-adbe5780-687f-11eb-97f9-4dfa9b50ee59.jpg)
![image](https://user-images.githubusercontent.com/38748663/107118180-a5feb300-687f-11eb-84c9-7d1dc9d74acc.png)

**Comment by ddobbelaere** [comment]
@vondele Maybe it's also interesting to try a run without gradient centralization? Maybe king weights from other neurons are inducing this behavior.

**Comment by vondele** [comment]
OK, worth looking into.. I've put that on the list.

**Comment by glinscott** [comment]
I tried a run with `use_gc=False` for Ranger, and it didn't do great, midpoint of run was much worse than typical:
```
Best nets so far:
   logs/nn-epoch142.nnue : -78.2 +- 13.5
   logs/nn-epoch184.nnue : -78.5 +- 26.9
   logs/nn-epoch182.nnue : -79.7 +- 16.1
```

**Comment by ddobbelaere** [comment]
@glinscott That's interesting. Do you mind sharing the nets of epoch 182 and 184? I'd love to do some analysis on those.

**Comment by glinscott** [comment]
@ddobbelaere Ah, unfortunately, I already deleted that run.  Will try and fit in another one with a different LR to see if things are any better.

**Comment by vondele** [comment]
And I have some interesting results with the factorizer off... First, while the optimization starts with worse nets, actually it catches up quickly.. I'm only at epoch 83, but it looks pretty reasonable (-34Elo) (black with factorizer, red without)
![image](https://user-images.githubusercontent.com/4202567/107123821-203f2f80-68a0-11eb-97e1-bf196526c5df.png)

The net however, looks visually *very* different:
![image](https://user-images.githubusercontent.com/4202567/107123852-4b298380-68a0-11eb-9512-6c3263ee09c3.png)

I've uploaded a copy of the net here:
https://drive.google.com/file/d/1foKqH5-Uc82DwfeGuFgDcJ_alcQGOnKB/view?usp=sharing

**Comment by vondele** [comment]
Also, training (in fens/second) is significantly faster (in fens/s), so even if still a bit lower, it is better than a similar net trained for the same time. I'll continue the training on this one to see where it ends up.

**Comment by ddobbelaere** [comment]
@vondele Reordering ("swapping") king and piece indices yields the following picture (I actually argue we should make this ordering the default, because it's significantly more helpful, at least in investigating this issue...):

![reordered nn-factorizer-off-epoch83 nnue_input-weights](https://user-images.githubusercontent.com/38748663/107124806-3bad3900-68a6-11eb-855b-f3f8a7c640e6.jpg)
![image](https://user-images.githubusercontent.com/38748663/107124864-77e09980-68a6-11eb-86c6-e42433f61d0a.png)

I don't really know what to make of this. I suspect the update of the weights for "unlikely" king positions (higher up the board) lags behind more likely king positions, as this is part of the algorithm (they have a larger contribution the stochastic gradient by definition). At some point (for some biases), I suspect some king positions are "cut off" (don't update anymore) because of vanishing gradients.

**Comment by sergiovieri** [comment]
This is also trained with pytorch trainer but using SGD https://www.comp.nus.edu.sg/~sergio-v/nnue-pytorch/nn.nnue

Still at lr=0.5, but net roughly looks the same as a fully trained one. Actually how do you extract factorizer weights from .nnue?

**Comment by ddobbelaere** [comment]
> Actually how do you extract factorizer weights from .nnue?

You can't, as all virtual features are coalesced ("squashed") into real ones upon serialization to nnue.

**Comment by vondele** [comment]
> Actually how do you extract factorizer weights from .nnue?

not sure if that refers to my earlier comment, I've just trained a net with the argument `--features=HalfKP`, which doesn't enable the factorizer during training.

**Comment by ddobbelaere** [comment]
This is probably more some kind of research topic and not an issue per se :)
Also, it's not clear it still pertains to the HalfKAv2 architecture and current way of training.
